### ISSUE TITLE
Nested templates regression since 5.3

### ISSUE THREAD

**Issue by vudaltsov** [vimeo/psalm#9241]
Here's what Psalm v5.2 correctly reports for https://psalm.dev/r/af716b51e1:
```
ERROR: InvalidReturnType - .stuff/test.php:37:44 - The declared return type 'int' for IntValueProcessor::process is incorrect, got '''' (see https://psalm.dev/011)
    public function process(Value $value): mixed


ERROR: InvalidReturnStatement - .stuff/test.php:39:16 - The inferred type '''' does not match the declared return type 'int' for IntValueProcessor::process (see https://psalm.dev/128)
        return '';
```

Psalm v5.3 reports `No issues!`

**Comment by vudaltsov** [comment]
Interestingly the problem disappears if I remove `: mixed` return type in `IntValueProcessor::process`.

https://psalm.dev/r/1a20f5a3a4

**Comment by vudaltsov** [comment]
Regression occurred after #8926 was merged.

**Comment by vudaltsov** [comment]
Closing in favor of #9245 which has a simpler example.
